### ISSUE TITLE
"no instance of function template" occurs when using chrono

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#1753]
Is anyone else hitting this?

```#include <chrono>
using namespace std;  
class foo
{public:
	int func()
	{
			chrono::time_point<chrono::system_clock> now = chrono::system_clock::now();
			chrono::time_point<chrono::system_clock> now2 = chrono::system_clock::now();
			auto ms = chrono::duration_cast<chrono::milliseconds>(now2 - now);
                        return 0;
	}
}
```

Error is 'no instance of function template "std::chrono::duration_cast" matches the argument list -- argument types are: (\<error-type\>)', with hovering over "now" showing an <error-type> in the 2nd template arg.

Repros with clang-x64 mode with clang and gcc 6 headers (on all OS's).

**Comment by sean-mcmanus** [comment]
Adding -stdlib=libc++ to the compilerPath fixed this for me, so I'm going with "by design" for now. So if anyone is using that to compile with, they should be sure to add it -- we don't automatically add that from compile_commands.json yet.

**Comment by riccozzz** [comment]
I am having this same issue on windows using mingw-w64. It was also happening on my linux machine, however adding -stdlib=libc++ to the compilerPath fixed the issue. Unsurprisingly doesn't work on my windows machine, probably because libc++ does not exist with mingw.

This seems like a bit of a bug anyways, no?

**Comment by sean-mcmanus** [comment]
Okay, yeah, I'm able to repro the issue with mingw-64. The problem is in order to get the VS compiler guys to fix this I have to provide them with a "minimum repro" that doesn't have any dependencies on library, which I attempted to do previously, but it was not an easy task given the dependencies chrono has on other headers like `<ratio>`. So I can't make any ETA guarantees (but I think we can investigate in May after 0.17.0 is shipped). If anyone has time to do this "reduction" that would enable us to get a fix sooner -- it would help a lot if we had a tool that simplifies the code to the minimum required for compilation.

**Comment by riccozzz** [comment]
Well, I got it to stop complaining by changing these two lines of code (in a class declaration, .hpp file):

```
using hrclk_t = std::chrono::high_resolution_clock;
using timepoint_t = std::chrono::time_point<hrclk_t>;
```
To:
```
using hrclk_t = std::chrono::high_resolution_clock;
using timepoint_t = std::chrono::high_resolution_clock::time_point;
```

Note that it stopped it from complaining _everywhere_ after I did that, not just within that class. I was confused because I had been using duration_cast before without issue and it just started to happen recently.

Basically, std::chrono::time_point\<foo\> breaks stuff, using std::chrono::foo::timepoint instead doesn't break stuff. Working example of code in OP:

```
using namespace std;
class foo {
  public:
    int func() {
        chrono::system_clock::time_point now = chrono::system_clock::now();
        chrono::system_clock::time_point now2 = chrono::system_clock::now();
        auto ms = chrono::duration_cast<chrono::milliseconds>(now2 - now);
        return 0;
    }
};
```

**Comment by sean-mcmanus** [comment]
Yeah, simply using code `chrono::system_clock a;` also gets rid of the error. Strange. We want to get this fixed eventually, because it could be a widespread issue with how we're parsing stuff.

**Comment by sean-mcmanus** [comment]
A related VS issue might be https://developercommunity.visualstudio.com/content/problem/278730/intellisense-only-issue-no-instance-of-function-te-1.html . Repros with functional. No ETA on when this will get fixed yet.

**Comment by sean-mcmanus** [comment]
Fixed with 0.18.0.
